toolchain/lex/tokenized_buffer.h

```diff
@@ -400,13 +397,12 @@ class TokenizedBuffer : public Printable<TokenizedBuffer> {
     // Payload values are typically ID types for which we create at most one per
     // token, so we ensure that `token_payload_` is large enough to fit any
     // token index. Stores to this field may overflow, but we produce an error
-    // in `Lexer::Finalize` if the file has more than `MaxTokens` tokens, so
-    // this value never overflows if lexing succeeds.
-    TokenKind::RawEnumType kind_ : sizeof(TokenKind) * 8;
+    // in `Lexer::Finalize` if the file has more than `TokenIndex::Max` tokens,
+    // so this value never overflows if lexing succeeds.
+    TokenKind kind_;
+    static_assert(sizeof(kind_) == 1, "TokenKind must pack to 8 bits");
```

Comment:
I think this is changing more than maybe you intended?

Previously, all three fields are part of a single bitfield, allowing it to be loaded and stored as a single memory access. With this change, the kind is a separate field. I think this may make it more expensive to update the `has_leading_space_` and `token_payload_`, as writes to those cannot write to `kind_`.

Reply:
Well... This *does* change quite a lot...

The generated code in the lexer is very different after this change for example. But the change completely baffles me. It is *significantly* better with this change.

Previously, we would generate code sequences like on AArch64:

```
       cmp   w25, #0x0
       mov   w9, #0x72                       // TokenKind 0x72
       mov   w10, #0x172                     // TokenKind 0x72 | (1 << 8)
       csel  w9, w10, w9, ne // ne = any
       ldr   x10, [x19, #112]
       adrp  x11, _DYNAMIC+0xdb60
       orr   w9, w9, w24, lsl #9
       ldr   x11, [x11, #352]
       orr   x9, x9, x23, lsl #32
       str   x9, [x10, w8, uxtw #3]
```

This is pre-merging the constant token kind in this code path with the flag for leading space, then selecting based on whether the flag was set, and then combining the rest... I... I have no idea why the previous bitfield oriented code produces this bizarre generated code.

With this PR, we get much nicer:

```
       lsl   x9, x23, #32
       bfi   x24, x25, #1, #23
       ldr   x10, [x19, #112]
       adrp  x12, _DYNAMIC+0xd9b8
       mov   w11, #0x72                      // TokenKind 0x72
       orr   x9, x9, x24, lsl #8
       ldr   x12, [x12, #2552]
       orr   x9, x9, x11
       str   x9, [x10, w8, uxtw #3]
```

Here, `x24` has the value of `has_leading_space_`, either `0` or `1`. Then we do `bfi` or "bitfield insert", to put the payload from bits 1-23. And then when we `or` it with the byte index we do `lsl #8` to shift it up by 8 more creating space for the token kind, and or-ing that kind in.

This is quite a bit better code. One instruction fewer, two micro-ops fewer by my reading, etc.

So, I'm quite surprised that the code intending to make it easy for LLVM to emit this well backfires, and this code which is a bit trickier for the optimizer still ends up better... But it does! So, yay?

This is a measurable difference in the lexer, about 1.3% on my M1.

Anyways, ignore my original comment on this PR. This change at least seems good. I may send a follow-up to add a comment to the code that surfaces how fragile the generated code quality is here and what kinds of options we've considered.

Reply:
I feel like I stumbled on something here that yes, is not what I intended. I'd mostly thought the removal of the `TokenKind::Make` call would be a small improvement for readability, but overall a noop due to the equivalent packing. But I'd failed to consider consequences from the type boundary.

Is it possible that the x86 code is worse? I've accepted the suggestion, but if the expectation is that this is worse, should I revert it back? Should I ask which version of LLVM you test with?

Reply:
To be clear, the `emplace_back` and the bitfield thing are separate.

`push_back` is substantially better than `emplace_back` on both x86 and arm.

For this issue, the bitfields, x86 shows no difference between the old code and the new code, but arm is much better with the new code. While it's surprising, we should keep the good code on arm, and that's by keeping the code as-is. I can send a PR to add some comments to the code for future updates to this.

But basically, this PR I think is good-to-go now.